.github/scripts/predictive-test-selection/README.md

```diff
@@ -0,0 +1,97 @@
+
+**Training mode (multiple PRs):**
+
+1. Retrieves the `N * PR_OVERSAMPLE_MULTIPLIER` most recently created PRs, oversampling to account for PRs that lack test runs or Cucumber reports.
+2. For each of the `N` PRs:
+    - Finds the **first completed** test run during the PR's lifetime that includes Cucumber reports (if none exist, the PR is skipped).
+    - Downloads all Cucumber JSON reports from this initial test run.
+    - Extracts and retains only secondary/recommended Cucumber test reports.
+    - Identifies the files modified that triggered this test run.
+    - Extracts the unique file extensions of the modified files.
+    - Retrieves the change history for those files over several recent time windows.
+    - Outputs a CSV file containing PR features: file extensions, change history, and PR number.
+
+### Q&A
+
+#### Why extract these specific PR features?
+
+These features are inspired by the [Facebook paper](https://arxiv.org/pdf/1810.05286), where file extensions can indicate which types of changes are more likely to cause test failures (e.g., editing a `.java` vs. a `.xml` file), and file change history highlights active areas of development that are more prone to breakages.
+
+While file extensions were useful at Facebook, they may be less informative for Uyuni, which is predominantly Java. Therefore, we plan to experiment with additional PR features as discussed in [#10553](https://github.com/uyuni-project/uyuni/issues/10553).
+
+#### Why use the first test run with Cucumber reports, not the last?
+
+My intuition is that we should focus on the first test run with Cucumber reports because it best reflects the real-world scenario we want to model: predicting test failures as soon as a new PR is opened, before any feedback or fixes have occurred. By using the initial test run, we capture the state of the code as it was originally submitted, ensuring our predictions are based on the same conditions developers face when their code is first tested. This approach avoids any bias introduced by subsequent test runs, which may include fixes or changes made in response to earlier failures.
+
+#### Why only keep secondary/recommended Cucumber test reports?
+
+Since the other tests are core tests, they are always executed at the start of every test run and cannot be skipped. They are mandatory for all PRs. Since these tests will always run regardless of our predictions, analyzing their results does not help us optimize test selection. Our focus is on secondary tests, which can be selectively executed.
+
+#### Will the script get rate-limited by the GitHub API given a large N?
+
+In practice, running the script with `N = 500` allowed us to collect data for about 150 PRs. We did not hit GitHub API rate limits, the script simply could not obtain more data because older PRs did not have stored Cucumber reports. The script continues processing until it gathers data for `N` PRs or exhausts all `N * PR_OVERSAMPLE_MULTIPLIER` candidates.
```

Comment:
Not critical while we run it manually, but if we make a GH action to run this script, I think we need to set some timeout too. Still, we can handle the timeout from outside the script.